tests/unit/test_contracts/bls_primitives_tests.cpp

```diff
@@ -203,118 +182,89 @@ class [[eosio::contract]] bls_primitives_tests : public contract{
-            check(0 == memcmp(r, res.data(), sizeof(bls_g2)), "bls_g2_exp test failed");
+            bls_g2_exp(reinterpret_cast<const bls_g2*>(&points[0]), reinterpret_cast<const bls_scalar*>(&scalars[0]), num, r);
+            check(std::equal(res.begin(), res.end(), r.begin()), "bls_g2_exp test failed");
         }
 
         [[eosio::action]]
-        void testpairing(const std::vector<uint8_t>& g1_points, const std::vector<uint8_t>& g2_points, const std::vector<uint8_t>& res)
-        {
-            check(g1_points.size()/sizeof(bls_g1) == g2_points.size()/sizeof(bls_g2), "number of elements in g1_points and g2_points must be equal");
-            uint32_t num = g1_points.size()/sizeof(bls_g1);
-            const bls_g1* pp1 = reinterpret_cast<const bls_g1*>(g1_points.data());
-            const bls_g2* pp2 = reinterpret_cast<const bls_g2*>(g2_points.data());
+        void testpairing(const std::vector<char>& g1_points, const std::vector<char>& g2_points, const std::vector<char>& res) {
+            auto num = g2_points.size()/sizeof(bls_g2);
+            check(g1_points.size()/sizeof(bls_g1) == num, "number of elements in g1_points and g2_points must be equal");
             bls_gt r;
-            bls_pairing(pp1, pp2, num, r);
-            check(0 == memcmp(r, res.data(), sizeof(bls_gt)), "bls_pairing test failed");
+            bls_pairing(reinterpret_cast<const bls_g1*>(&g1_points[0]), reinterpret_cast<const bls_g2*>(&g2_points[0]), num, r);
+            check(std::equal(res.begin(), res.end(), r.begin()), "bls_pairing test failed");
         }
 
         [[eosio::action]]
-        void testg1map(const std::vector<uint8_t>& e, const std::vector<uint8_t>& res)
-        {
-            bls_fp element;
-            memcpy(element, e.data(), sizeof(bls_fp));
+        void testg1map(const std::vector<char>& e, const std::vector<char>& res) {
             bls_g1 r;
-            bls_g1_map(element, r);
-            check(0 == memcmp(r, res.data(), sizeof(bls_g1)), "bls_g1_map test failed");
+            bls_g1_map(reinterpret_cast<const bls_fp&>(e[0]), r);
+            check(std::equal(res.begin(), res.end(), r.begin()), "bls_g1_map test failed");
         }
 
         [[eosio::action]]
-        void testg2map(const std::vector<uint8_t>& e, const std::vector<uint8_t>& res)
-        {
-            bls_fp2 element;
-            memcpy(element, e.data(), sizeof(bls_fp2));
+        void testg2map(const std::vector<char>& e, const std::vector<char>& res) {
             bls_g2 r;
-            bls_g2_map(element, r);
-            check(0 == memcmp(r, res.data(), sizeof(bls_g2)), "bls_g2_map test failed");
+            bls_g2_map(reinterpret_cast<const bls_fp2&>(e[0]), r);
+            check(std::equal(res.begin(), res.end(), r.begin()), "bls_g2_map test failed");
         }
 
         // Construct an extensible-output function based on SHA256
         void xmd_sh256(
-            uint8_t *buf,
+            char *buf,
```

Comment:
I'm not sure why the change from `uint8_t` to `char`, but if we need to change I'd prefer `std::byte` unless there is a reason `char` is needed.

Reply:
the thing is, we have char inside low level interfaces. so we have to cast it on some level. I'm trying to reduce that.
https://github.com/AntelopeIO/cdt/blob/bls_integration/libraries/eosiolib/core/eosio/crypto_bls_ext.hpp